### PR TITLE
Shellcheck: ci/

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: Shellcheck: ci/
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue] Partial fix for #7793 
[Optional Deprecated:]

### CODE REVIEW

File: ci/docker_ci.sh [hunk 1]
```diff
@@ -50,37 +50,39 @@ use_builder() {
...
 
 build_images() {
+  local args
   TYPE=$1
   BUILD_TAG=$2
 
   use_builder "${TYPE}"
-  ARGS="$(build_args ${TYPE})"
-  PLATFORM="$(build_platforms ${TYPE})"
+  IFS=' ' read -ra args < <(build_args "${TYPE}")
```
Comment (phlax) [inline ci/docker_ci.sh:58]: use `while read`

File: ci/docker_ci.sh [hunk 2]
```diff
@@ -50,37 +50,39 @@ use_builder() {
...
 
 push_images() {
+  local args
   TYPE=$1
   BUILD_TAG=$2
 
   use_builder "${TYPE}"
-  ARGS="$(build_args ${TYPE})"
-  PLATFORM="$(build_platforms ${TYPE})"
+  IFS=' ' read -ra args < <(build_args "${TYPE}")
```
Comment (phlax) [inline ci/docker_ci.sh:82]: use `while read`

File: ci/mac_ci_steps.sh [hunk 3]
```diff
@@ -11,13 +11,22 @@ trap finish EXIT
...
-  --action_env=PATH=/usr/local/bin:/opt/local/bin:/usr/bin:/bin --test_output=all \
-  --flaky_test_attempts=integration@2 ${BAZEL_BUILD_EXTRA_OPTIONS} ${BAZEL_EXTRA_TEST_OPTIONS}"
+BAZEL_BUILD_OPTIONS=(
+    "--curses=no"
+    "--show_task_finish"
+    "--verbose_failures"
+    "--action_env=PATH=/usr/local/bin:/opt/local/bin:/usr/bin:/bin"
+    "--test_output=all"
+    "--flaky_test_attempts=integration@2")
+readarray -t -O "${#BAZEL_BUILD_OPTIONS[@]}" BAZEL_BUILD_OPTIONS \
```
Comment (phlax) [inline ci/mac_ci_steps.sh:26]: use `read while`

File: ci/run_envoy_docker.sh [hunk 4]
```diff
@@ -2,7 +2,15 @@
 
 set -e
 
-. $(dirname $0)/envoy_build_sha.sh
+# shellcheck source=ci/envoy_build_sha.sh
+. "$(dirname "$0")"/envoy_build_sha.sh
+
+IFS=' ' read -ra ENVOY_DOCKER_OPTIONS < <(echo "${ENVOY_DOCKER_OPTIONS:-}")
```
Comment (phlax) [inline ci/run_envoy_docker.sh:8]: -`IFS` and simplify `<<<`

File: ci/windows_ci_steps.sh [hunk 5]
```diff
@@ -19,30 +20,36 @@ df -h
...
 # Default to msvc-cl if not overridden
-BAZEL_BUILD_EXTRA_OPTIONS=${BAZEL_BUILD_EXTRA_OPTIONS:---config=msvc-cl}
-BAZEL_BUILD_OPTIONS="-c opt --show_task_finish --verbose_failures \
-  --test_output=errors ${BAZEL_BUILD_EXTRA_OPTIONS} ${BAZEL_EXTRA_TEST_OPTIONS}"
+BAZEL_BUILD_OPTIONS=(
+    "-c opt"
+    "--show_task_finish"
+    "--verbose_failures"
+    "--test_output=errors")
+readarray -t -O "${#BAZEL_BUILD_OPTIONS[@]}" BAZEL_BUILD_OPTIONS \
```
Comment (phlax) [inline ci/windows_ci_steps.sh:35]: simplify this and remove `readarray`

File: ci/do_circle_ci.sh [hunk 6]
```diff
@@ -30,8 +31,13 @@ export NUM_CPUS=6
 
 # CircleCI doesn't support IPv6 by default, so we run all tests with IPv4 only.
 # IPv6 tests are run with Azure Pipelines.
-export BAZEL_BUILD_EXTRA_OPTIONS+="--test_env=ENVOY_IP_TEST_VERSIONS=v4only --local_cpu_resources=${NUM_CPUS} \
-  --action_env=HOME --action_env=PYTHONUSERBASE --test_env=HOME --test_env=PYTHONUSERBASE"
+export BAZEL_BUILD_EXTRA_OPTIONS=" \
```
Comment (phlax) [inline ci/do_circle_ci.sh:34]: retain the `+=`

File: ci/windows_ci_steps.sh [hunk 7]
```diff
@@ -19,30 +20,34 @@ df -h
...
 # Default to msvc-cl if not overridden
-BAZEL_BUILD_EXTRA_OPTIONS=${BAZEL_BUILD_EXTRA_OPTIONS:---config=msvc-cl}
-BAZEL_BUILD_OPTIONS="-c opt --show_task_finish --verbose_failures \
-  --test_output=errors ${BAZEL_BUILD_EXTRA_OPTIONS} ${BAZEL_EXTRA_TEST_OPTIONS}"
+BAZEL_BUILD_OPTIONS=(
+    "-c opt"
+    "--show_task_finish"
+    "--verbose_failures"
+    "--test_output=errors"
+    "${BAZEL_BUILD_EXTRA_OPTIONS:---config=msvc-cl}"
```
Comment (phlax) [inline ci/windows_ci_steps.sh:35]: split to array

File: ci/run_envoy_docker_windows.sh [hunk 8]
```diff
@@ -19,9 +23,9 @@ ENVOY_SOURCE_DIR=$(echo "${PWD}" | sed -E "s#/([a-zA-Z])/#\1:/#")
 export ENVOY_BUILD_IMAGE="${IMAGE_NAME}:${IMAGE_ID}"
 
 # Since we specify an explicit hash, docker-run will pull from the remote repo if missing.
-docker run --rm ${DOCKER_TTY_OPTION} -e HTTP_PROXY=${http_proxy} -e HTTPS_PROXY=${https_proxy} \
-  ${GIT_VOLUME_OPTION} -e BAZEL_BUILD_EXTRA_OPTIONS -e BAZEL_EXTRA_TEST_OPTIONS -e BAZEL_REMOTE_CACHE \
+docker run --rm "${DOCKER_TTY_OPTION}" -e HTTP_PROXY="${http_proxy}" -e HTTPS_PROXY="${https_proxy}" \
```
Comment (phlax) [inline ci/run_envoy_docker_windows.sh:26]: split lines

File: ci/run_envoy_docker_windows.sh [hunk 9]
```diff
@@ -3,7 +3,11 @@
 set -e
 
 # The image tag for the Windows image is the same as the Linux one so we use the same mechanism to find it
-. $(dirname $0)/envoy_build_sha.sh
+# shellcheck source=ci/envoy_build_sha.sh
+. "$(dirname "$0")/envoy_build_sha.sh"
+
+http_proxy="${http_proxy:-}"
+https_proxy="${https_proxy:-}"
```
Comment (phlax) [inline ci/run_envoy_docker_windows.sh:10]: switch case here and simplify docker args

File: ci/docker-entrypoint.sh [hunk 10]
```diff
@@ -1,6 +1,9 @@
 #!/usr/bin/env sh
 set -e
 
+loglevel="${loglevel:-}"
+
+
```
Comment (zuercher) [inline ci/docker-entrypoint.sh:6]: nit: rm extra blank line

File: ci/run_envoy_docker.sh [hunk 11]
```diff
@@ -2,7 +2,15 @@
...
+
+# these ~globals should be uppercase
```
Comment (zuercher) [inline ci/run_envoy_docker.sh:10]: Make this `# TODO(phlax): ...` ?

File: ci/envoy_build_sha.sh [hunk 12]
```diff
@@ -1,2 +1,5 @@
-ENVOY_BUILD_SHA=$(grep envoyproxy/envoy-build-ubuntu $(dirname $0)/../.bazelrc | sed -e 's#.*envoyproxy/envoy-build-ubuntu:\(.*\)#\1#' | uniq)
+#!/bin/bash
+
+
```
Comment (zuercher) [inline ci/envoy_build_sha.sh:3]: nit: rm extra blank line

**Review by zuercher — COMMENTED** [review]
Looks good. Just some small nits.

File: ci/do_circle_ci.sh (hunk 6 above)
Comment (zuercher) [reply ci/do_circle_ci.sh:34]: I agree that we need to keep the append.

File: ci/build_setup.sh [hunk 13]
```diff
@@ -89,7 +89,6 @@ export BAZEL_QUERY_OPTIONS="${BAZEL_OPTIONS}"
 # Use https://docs.bazel.build/versions/master/command-line-reference.html#flag--experimental_repository_cache_hardlinks
 # to save disk space.
 export BAZEL_BUILD_OPTIONS=(
-  "${BAZEL_OPTIONS:-}"
```
Comment (phlax) [inline ci/build_setup.sh]: readd  as array

File: ci/build_setup.sh [hunk 14]
```diff
@@ -6,17 +6,24 @@ set -e
...
 [ -z "${ENVOY_SRCDIR}" ] && export ENVOY_SRCDIR=/source
 [ -z "${ENVOY_BUILD_TARGET}" ] && export ENVOY_BUILD_TARGET=//source/exe:envoy-static
 [ -z "${ENVOY_BUILD_DEBUG_INFORMATION}" ] && export ENVOY_BUILD_DEBUG_INFORMATION=//source/exe:envoy-static.dwp
-[ -z "${ENVOY_BUILD_ARCH}" ] && export ENVOY_BUILD_ARCH=$(uname -m)
+[ -z "${ENVOY_BUILD_ARCH}" ] && {
+    ENVOY_BUILD_ARCH=$(uname -m)
+    export ENVOY_BUILD_ARCH
+}
+
+IFS=' ' read -ra BAZEL_BUILD_EXTRA_OPTIONS < <(echo "${BAZEL_BUILD_EXTRA_OPTIONS:-}")
```
Comment (phlax) [inline ci/build_setup.sh:18]: rm `IFS` and simplify

**Comment by phlax** [comment]
This should be ready for review.

Apologies for the rebasing/squashing/force-pushing - i needed to test  out quite a few things

File: ci/mac_ci_steps.sh [hunk 15]
```diff
@@ -11,13 +11,25 @@ trap finish EXIT
...
+read -ra BAZEL_BUILD_EXTRA_OPTIONS <<< "${BAZEL_BUILD_EXTRA_OPTIONS:-}"
+read -ra BAZEL_EXTRA_TEST_OPTIONS <<< "${BAZEL_EXTRA_TEST_OPTIONS:-}"
+
 # TODO(zuercher): remove --flaky_test_attempts when https://github.com/envoyproxy/envoy/issues/2428
 # is resolved.
-BAZEL_BUILD_OPTIONS="--curses=no --show_task_finish --verbose_failures \
-  --action_env=PATH=/usr/local/bin:/opt/local/bin:/usr/bin:/bin --test_output=all \
-  --flaky_test_attempts=integration@2 ${BAZEL_BUILD_EXTRA_OPTIONS} ${BAZEL_EXTRA_TEST_OPTIONS}"
+BAZEL_BUILD_OPTIONS=(
+    "${BAZEL_BUILD_OPTIONS[@]}"
```
Comment (phlax) [inline ci/mac_ci_steps.sh:24]: remove BAZEL_BUILD_OPTIONS

File: ci/windows_ci_steps.sh [hunk 16]
```diff
@@ -11,38 +11,48 @@ trap finish EXIT
...
   /c/windows/system32/cmd.exe /c "mklink /d $drive:\\$drive $drive:\\"
 fi
 
 BAZEL_STARTUP_OPTIONS="--output_base=c:/_eb"
 # Default to msvc-cl if not overridden
-BAZEL_BUILD_EXTRA_OPTIONS=${BAZEL_BUILD_EXTRA_OPTIONS:---config=msvc-cl}
-BAZEL_BUILD_OPTIONS="-c opt --show_task_finish --verbose_failures \
-  --test_output=errors ${BAZEL_BUILD_EXTRA_OPTIONS} ${BAZEL_EXTRA_TEST_OPTIONS}"
+BAZEL_BUILD_OPTIONS=(
+    "${BAZEL_BUILD_OPTIONS[@]}"
```
Comment (phlax) [inline ci/windows_ci_steps.sh:35]: rm BAZEL_BUILD_OPTIONS

File: ci/do_ci.sh [hunk 17]
```diff
@@ -405,8 +414,10 @@ elif [[ "$CI_TARGET" == "docs" ]]; then
 elif [[ "$CI_TARGET" == "verify_examples" ]]; then
   echo "verify examples..."
   docker load < "$ENVOY_DOCKER_BUILD_DIR/docker/envoy-docker-images.tar.xz"
-  images=($(docker image list --format "{{.Repository}}"))
-  tags=($(docker image list --format "{{.Tag}}"))
+  readarray -t images \
```
Comment (phlax) [inline ci/do_ci.sh:417]: replace `readarray`

**Review by zuercher — APPROVED** [review]
Thanks for taking this on! I had one question for my own edification.

File: ci/do_ci.sh [hunk 18]
```diff
@@ -93,6 +95,13 @@ function cp_binary_for_image_build() {
       bazel-bin/"${ENVOY_BIN}"{,.dwp}
 }
 
+function bazel_build () {
+    # build options are turned to string before building
+    # echo "BAZEL BUILD: ${BAZEL_BUILD_OPTIONS[@]/#/X}"
+    _build=(build "${BAZEL_BUILD_OPTIONS[@]}" "$@")
+    bazel "${_build[@]}"
+}
```
[the remark this answers is not in the document]
Comment (phlax) [reply ci/do_ci.sh:103]: ill factor it back out - i was debugging a test

**Review by zuercher — DISMISSED** [review]
Thanks!

**Comment by mattklein123** [comment]
Needs main merge, sorry.

/wait

**Review by zuercher — APPROVED** [review]

**Pull request merged**